File: app/scripts/resources/scripts/call_forward.lua
--
--	FusionPBX
--	Version: MPL 1.1
--
--	The contents of this file are subject to the Mozilla Public License Version
--	1.1 (the "License"); you may not use this file except in compliance with
--	the License. You may obtain a copy of the License at
--	http://www.mozilla.org/MPL/
--
--	Software distributed under the License is distributed on an "AS IS" basis,
--	WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
--	for the specific language governing rights and limitations under the
--	License.
--
--	The Original Code is FusionPBX
--
--	The Initial Developer of the Original Code is
--	Mark J Crane <markjcrane@fusionpbx.com>
--	Copyright (C) 2010-2017
--	the Initial Developer. All Rights Reserved.
--
--	Contributor(s):
--	Mark J Crane <markjcrane@fusionpbx.com>

--set default variables
	min_digits = "1";
	max_digits = "17";
	max_tries = "3";
	digit_timeout = "5000";

--define the trim function
	require "resources.functions.trim"

--create the api object
	api = freeswitch.API();

--includes
	require "resources.functions.config";
	require "resources.functions.channel_utils";
	local log = require "resources.functions.log".call_forward
	local cache = require "resources.functions.cache"
	local Database = require "resources.functions.database"
	local Settings = require "resources.functions.lazy_settings"
	local blf = require "resources.functions.blf"
	local notify = require "app.feature_event.resources.functions.feature_event_notify"	

--include json library
	local json
	if (debug["sql"]) then
		json = require "resources.functions.lunajson"
	end

	local function empty(t)
		return (not t) or (#t == 0)
	end

--check if the session is ready
	if not session:ready() then return end

--answer the call
	session:answer();

--get the variables
	local enabled = session:getVariable("enabled");
	local pin_number = session:getVariable("pin_number");
	local sounds_dir = session:getVariable("sounds_dir");
	local domain_uuid = session:getVariable("domain_uuid");
	local domain_name = session:getVariable("domain_name");
	local extension_uuid = session:getVariable("extension_uuid");
	local request_id = session:getVariable("request_id");
	local forward_all_destination = session:getVariable("forward_all_destination") or '';
	local extension;

--set the sounds path for the language, dialect and voice
	local default_language = session:getVariable("default_language") or 'en';
	local default_dialect = session:getVariable("default_dialect") or 'us';
	local default_voice = session:getVariable("default_voice") or 'callie';

--a moment to sleep
	session:sleep(1000);

--connect to the database
	local dbh = Database.new('system');

	local settings = Settings.new(dbh, domain_name, domain_uuid);

--request id is true
	if (request_id == "true") then
		--unset extension uuid
			extension_uuid = nil;

		--get the extension
			if not session:ready() then return end
			local min_digits = 2;
			local max_digits = 20;
			extension = session:playAndGetDigits(min_digits, max_digits, max_tries, digit_timeout, "#", "phrase:voicemail_enter_id:#", "", "\\d+");
			if empty(extension) then return end

		--get the pin number
			if not session:ready() then return end
			min_digits = 3;
			max_digits = 20;
			local caller_pin_number = session:playAndGetDigits(min_digits, max_digits, max_tries, digit_timeout, "#", "phrase:voicemail_enter_pass:#", "", "\\d+");
			if empty(caller_pin_number) then return end

		--check to see if the pin number is correct
			if not session:ready() then return end
			local sql = "SELECT voicemail_password FROM v_voicemails ";
			sql = sql .. "WHERE domain_uuid = :domain_uuid ";
			sql = sql .. "AND voicemail_id = :extension ";
			local params = {domain_uuid = domain_uuid, extension = extension};
			if (debug["sql"]) then
				log.noticef("SQL: %s; params: %s", sql, json.encode(params));
			end
			local voicemail_password = dbh:first_value(sql, params)
			if (voicemail_password ~= caller_pin_number) then
				--access denied
				session:streamFile("phrase:voicemail_fail_auth:#");
				return session:hangup("NORMAL_CLEARING");
			end
	end

--determine whether to update the dial string
	if not session:ready() then return end

	local sql = "select * from v_extensions ";
	sql = sql .. "where domain_uuid = :domain_uuid ";
	local params = {domain_uuid = domain_uuid};
	if (extension_uuid ~= nil) then
		sql = sql .. "and extension_uuid = :extension_uuid ";
		params.extension_uuid = extension_uuid;
	else
		sql = sql .. "and (extension = :extension or number_alias = :extension) ";
		params.extension = extension;
	end
	if (debug["sql"]) then
		log.noticef("SQL: %s; params: %s", sql, json.encode(params));
	end
	local row = dbh:first_row(sql, params)
	if not row then return end

	extension_uuid = row.extension_uuid;
	extension = row.extension;
	local number_alias = row.number_alias or '';
	local accountcode = row.accountcode;
	local forward_all_enabled = row.forward_all_enabled;
	local last_forward_all_destination = row.forward_all_destination;
	local follow_me_uuid = row.follow_me_uuid;
	local toll_allow = row.toll_allow or '';

--toggle enabled
	if enabled == "toggle" then
		-- if we toggle CF and specify new destination number then just enable it
		if (#forward_all_destination == 0) or (forward_all_destination == row.forward_all_destination) then
			enabled = (forward_all_enabled == "true") and "false" or "true";
		else
			enabled = 'true'
		end
	end

-- get destination number form database if it not provided
	if enabled == 'true' and #forward_all_enabled == 0 then
		forward_all_destination = row.forward_all_destination
	end

	if not session:ready() then return end

--get the destination by argument and set the forwarding destination 
	destination_by_arg = argv[1];
	
	if enabled == "true" and destination_by_arg then
		forward_all_destination = destination_by_arg;
	end
	
--get the forward destination by IVR if destination has not been passed by argument
	if enabled == "true" and empty(forward_all_destination) and not destination_by_arg then
		forward_all_destination = session:playAndGetDigits(min_digits, max_digits, max_tries, digit_timeout, "#", sounds_dir.."/"..default_language.."/"..default_dialect.."/"..default_voice.."/ivr/ivr-enter_destination_telephone_number.wav", "", "\\d+");
		if empty(forward_all_destination) then return end
	end

--set call forward
	if enabled == "true" then
		--set forward_all_enabled
			forward_all_enabled = "true";
			channel_display(session:get_uuid(), "Activated")
		--say the destination number
			session:say(forward_all_destination, default_language, "number", "iterated");
		--notify the caller
			session:streamFile(sounds_dir.."/"..default_language.."/"..default_dialect.."/"..default_voice.."/ivr/ivr-call_forwarding_has_been_set.wav");
	end

--get default caller_id for outbound call
	if enabled == "true" and settings:get('cdr', 'call_forward_fix', 'boolean') == 'true' then
		if not empty(row.outbound_caller_id_number) then
			forward_caller_id = forward_caller_id ..
				 ",outbound_caller_id_number=" .. row.outbound_caller_id_number ..
				 ",origination_caller_id_number=" .. row.outbound_caller_id_number
		end
		if not empty(row.outbound_caller_id_name) then
			forward_caller_id = forward_caller_id ..
				 ",outbound_caller_id_name=" .. row.outbound_caller_id_name ..
				 ",origination_caller_id_name=" .. row.outbound_caller_id_name
		end
	end

--unset call forward
	if session:ready() and enabled == "false" then
		--set forward_all_enabled
			forward_all_enabled = "false";
			channel_display(session:get_uuid(), "Cancelled")
		--notify the caller
			session:streamFile(sounds_dir.."/"..default_language.."/"..default_dialect.."/"..default_voice.."/ivr/ivr-call_forwarding_has_been_cancelled.wav");
	end

--disable the follow me
	if enabled == "true" and not empty(follow_me_uuid) then
		local sql = "update v_follow_me set ";
		sql = sql .. "follow_me_enabled = 'false' ";
		sql = sql .. "where domain_uuid = :domain_uuid ";
		sql = sql .. "and follow_me_uuid = :follow_me_uuid ";
		local params = {domain_uuid = domain_uuid, follow_me_uuid = follow_me_uuid};
		if (debug["sql"]) then
			log.noticef("SQL: %s; params: %s", sql, json.encode(params));
		end
		dbh:query(sql, params);
	end

--check the destination
	if empty(forward_all_destination) then
		enabled = "false";
		forward_all_enabled = "false";
	end

--update the extension
	do
		local sql = "update v_extensions set ";
		if (enabled == "true") then
			sql = sql .. "forward_all_destination = :forward_all_destination, ";
			sql = sql .. "do_not_disturb = 'false', ";
		else
			sql = sql .. "forward_all_destination = null, ";
		end
		sql = sql .. "forward_all_enabled = :forward_all_enabled ";
		sql = sql .. "where domain_uuid = :domain_uuid ";
		sql = sql .. "and extension_uuid = :extension_uuid ";
		local params = {
			forward_all_destination = forward_all_destination;
			forward_all_enabled = forward_all_enabled;
			domain_uuid = domain_uuid;
			extension_uuid = extension_uuid;
		}
		if (debug["sql"]) then
			log.noticef("SQL: %s; params: %s", sql, json.encode(params));
		end
		dbh:query(sql, params);
	end

--send notify to phone if feature sync is enabled
	if settings:get('device', 'feature_sync', 'boolean') == 'true' then
		-- Get values from the database
			do_not_disturb, forward_all_enabled, forward_all_destination, forward_busy_enabled, forward_busy_destination, forward_no_answer_enabled, forward_no_answer_destination, call_timeout = notify.get_db_values(extension, domain_name);

		-- Get the sip_profile
			if (extension ~= nil and domain_name ~= nil) then
				sip_profiles = notify.get_profiles(extension, domain_name);
			end

		if (sip_profiles ~= nil) then
				freeswitch.consoleLog("NOTICE", "[feature_event] SIP NOTIFY: CFWD set to "..forward_all_enabled.."\n");

			--Do Not Disturb
				notify.dnd(extension, domain_name, sip_profiles, do_not_disturb);

			--Forward all
				forward_immediate_enabled = forward_all_enabled;
				forward_immediate_destination = forward_all_destination;

				--workaround for freeswitch not sending NOTIFY when destination values are nil. Send 0.
					if (string.len(forward_immediate_destination) < 1) then 
						forward_immediate_destination = '0';
					end

				freeswitch.consoleLog("NOTICE", "[feature_event] forward_immediate_destination "..forward_immediate_destination.."\n");
				notify.forward_immediate(extension, domain_name, sip_profiles, forward_immediate_enabled, forward_immediate_destination);

			--Forward busy
				--workaround for freeswitch not sending NOTIFY when destination values are nil. Send 0.
					if (string.len(forward_busy_destination) < 1) then
						forward_busy_destination = '0';
					end

				freeswitch.consoleLog("NOTICE", "[feature_event] forward_busy_destination "..forward_busy_destination.."\n");
				notify.forward_busy(extension, domain_name, sip_profiles, forward_busy_enabled, forward_busy_destination);

			--Forward No Answer
				ring_count = math.ceil (call_timeout / 6);
				--workaround for freeswitch not sending NOTIFY when destination values are nil. Send 0.
					if (string.len(forward_no_answer_destination) < 1) then 
						forward_no_answer_destination = '0';
					end

				freeswitch.consoleLog("NOTICE", "[feature_event] forward_no_answer_destination "..forward_no_answer_destination.."\n");
				notify.forward_no_answer(extension, domain_name, sip_profiles, forward_no_answer_enabled, forward_no_answer_destination, ring_count);
		end
	end

--disconnect from database
	dbh:release();

--clear the cache
	if extension and #extension > 0 and cache.support() then
		cache.del("directory:"..extension.."@"..domain_name);
		if #number_alias > 0 then
			cache.del("directory:"..number_alias.."@"..domain_name);
		end
	end

--hangup
	if (session:ready()) then
		--wait for the file to be written before proceeding
			session:sleep(100);
		--end the call
			session:hangup();
	end

-- BLF for display CF status
	blf.forward(enabled == 'true', extension, number_alias, 
		last_forward_all_destination, forward_all_destination, domain_name);

-- turn off DND BLF
	if (enabled == 'true') then
		blf.dnd(false, extension, number_alias, domain_name);
	end

File: app/call_forward/resources/classes/call_forward.php
<?php

	/*
	  FusionPBX
	  Version: MPL 1.1

	  The contents of this file are subject to the Mozilla Public License Version
	  1.1 (the "License"); you may not use this file except in compliance with
	  the License. You may obtain a copy of the License at
	  http://www.mozilla.org/MPL/

	  Software distributed under the License is distributed on an "AS IS" basis,
	  WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
	  for the specific language governing rights and limitations under the
	  License.

	  The Original Code is FusionPBX

	  The Initial Developer of the Original Code is
	  Mark J Crane <markjcrane@fusionpbx.com>
	  Copyright (C) 2010 - 2023
	  All Rights Reserved.

	  Contributor(s):
	  Mark J Crane <markjcrane@fusionpbx.com>
	  Luis Daniel Lucio Quiroz <dlucio@okay.com.mx>
	  Errol Samuels <voiptology@gmail.com>

	 */

//define the call_forward class
	class call_forward {

		// set class constants
		const APP_NAME = 'calls';
		const APP_UUID = '19806921-e8ed-dcff-b325-dd3e5da4959d';
		const PERMISSION = 'call_forward';
		const LIST_PAGE = 'calls.php';
		const TABLE = 'extensions';
		const UUID_PREFIX = 'extension_';
		const TOGGLE_FIELD = 'forward_all_enabled';
		const TOGGLE_VALUES = ['true', 'false'];

		public $debug;
		public $domain_uuid;
		public $domain_name;
		public $extension_uuid;
		public $forward_all_destination;
		public $forward_all_enabled;
		public $accountcode;
		public $outbound_caller_id_name;
		public $outbound_caller_id_number;

		/**
		 * declare private variables
		 */
		private $extension;
		private $number_alias;
		private $toll_allow;

		public function set() {
			//create the database connection
			$database = new database;

			//determine whether to update the dial string
			$sql = "select * from v_extensions ";
			$sql .= "where domain_uuid = :domain_uuid ";
			$sql .= "and extension_uuid = :extension_uuid ";
			$parameters['domain_uuid'] = $this->domain_uuid;
			$parameters['extension_uuid'] = $this->extension_uuid;
			$row = $database->select($sql, $parameters, 'row');
			if (is_array($row) && @sizeof($row) != 0) {
				$this->extension = $row["extension"];
				$this->number_alias = $row["number_alias"];
				$this->accountcode = $row["accountcode"];
				$this->toll_allow = $row["toll_allow"];
				$this->outbound_caller_id_name = $row["outbound_caller_id_name"];
				$this->outbound_caller_id_number = $row["outbound_caller_id_number"];
			}
			unset($sql, $parameters, $row);

			//build extension update array
			$array['extensions'][0]['extension_uuid'] = $this->extension_uuid;
			$array['extensions'][0]['forward_all_destination'] = strlen($this->forward_all_destination) != 0 ? $this->forward_all_destination : null;
			if (empty($this->forward_all_destination) || $this->forward_all_enabled == "false") {
				$array['extensions'][0]['forward_all_enabled'] = 'false';
			} else {
				$array['extensions'][0]['forward_all_enabled'] = 'true';
			}

			//grant temporary permissions
			$p = new permissions;
			$p->add('extension_add', 'temp');

			//execute update
			$database->app_name = 'calls';
			$database->app_uuid = '19806921-e8ed-dcff-b325-dd3e5da4959d';
			$database->save($array);
			unset($array);

			//revoke temporary permissions
			$p->delete('extension_add', 'temp');

			//delete extension from the cache
			$cache = new cache;
			$cache->delete("directory:" . $this->extension . "@" . $this->domain_name);
			if (!empty($this->number_alias)) {
				$cache->delete("directory:" . $this->number_alias . "@" . $this->domain_name);
			}
		}

		/**
		 * Toggle an array of call_forward records
		 * @param array $records array of records to toggle
		 */
		public function toggle(array $records) {

			//validate the token
			$token = new token;
			if (!$token->validate($_SERVER['PHP_SELF'])) {
				message::add($text['message-invalid_token'], 'negative');
				header('Location: ' . self::LIST_PAGE);
				exit;
			}

			//validate there are records to process
			if (count($records) < 1) return;

			//check we have permission for this action
			if (permission_exists(self::PERMISSION)) {

				//create the database connection
				$database = new database;

				//add multi-lingual support
				$language = new text;
				$text = $language->get();

				
				// initialize an empty array
				$uuids = [];
				$extensions = [];
				
				//get current toggle state
				foreach ($records as $x => $record) {
					if (!empty($record['checked']) && $record['checked'] == 'true' && is_uuid($record['uuid'])) {
						$uuids[] = "'" . $record['uuid'] . "'";
					}
				}

				//toggle the checked records
				if (count($uuids) > 0) {
					$sql = "select " . self::UUID_PREFIX . "uuid as uuid, extension, number_alias, ";
					$sql .= "call_timeout, do_not_disturb, ";
					$sql .= "forward_all_enabled, forward_all_destination, ";
					$sql .= "forward_busy_enabled, forward_busy_destination, ";
					$sql .= "forward_no_answer_enabled, forward_no_answer_destination, ";
					$sql .= self::TOGGLE_FIELD . " as toggle, follow_me_uuid ";
					$sql .= "from v_" . self::TABLE . " ";
					$sql .= "where (domain_uuid = :domain_uuid or domain_uuid is null) ";
					$sql .= "and " . self::UUID_PREFIX . "uuid in (" . implode(', ', $uuids) . ") ";
					$parameters['domain_uuid'] = $_SESSION['domain_uuid'];
					$rows = $database->select($sql, $parameters, 'all');
					if (is_array($rows) && @sizeof($rows) != 0) {
						foreach ($rows as $row) {
							$extensions[$row['uuid']]['extension'] = $row['extension'];
							$extensions[$row['uuid']]['number_alias'] = $row['number_alias'];
							$extensions[$row['uuid']]['call_timeout'] = $row['call_timeout'];
							$extensions[$row['uuid']]['do_not_disturb'] = $row['do_not_disturb'];
							$extensions[$row['uuid']]['forward_all_enabled'] = $row['forward_all_enabled'];
							$extensions[$row['uuid']]['forward_all_destination'] = $row['forward_all_destination'];
							$extensions[$row['uuid']]['forward_busy_enabled'] = $row['forward_busy_enabled'];
							$extensions[$row['uuid']]['forward_busy_destination'] = $row['forward_busy_destination'];
							$extensions[$row['uuid']]['forward_no_answer_enabled'] = $row['forward_no_answer_enabled'];
							$extensions[$row['uuid']]['forward_no_answer_destination'] = $row['forward_no_answer_destination'];
							$extensions[$row['uuid']]['state'] = $row['toggle'];
							$extensions[$row['uuid']]['follow_me_uuid'] = $row['follow_me_uuid'];
						}
					}
					unset($sql, $parameters, $rows, $row);
				}

				//build update array
				$x = 0;
				foreach ($extensions as $uuid => $extension) {

					//check destination
					$destination_exists = $extension['forward_all_destination'] != '' ? true : false;

					//determine new state
					$new_state = $extension['state'] == self::TOGGLE_VALUES[1] && $destination_exists ? self::TOGGLE_VALUES[0] : self::TOGGLE_VALUES[1];

					//toggle feature
					if ($new_state != $extension['state']) {
						$array[self::TABLE][$x][self::UUID_PREFIX . 'uuid'] = $uuid;
						$array[self::TABLE][$x][self::TOGGLE_FIELD] = $new_state;
					}

					//disable other features
					if ($new_state == self::TOGGLE_VALUES[0]) { //true
						$array[self::TABLE][$x]['do_not_disturb'] = self::TOGGLE_VALUES[1]; //false
						$array[self::TABLE][$x]['follow_me_enabled'] = self::TOGGLE_VALUES[1]; //false
						if (is_uuid($extension['follow_me_uuid'])) {
							$array['follow_me'][$x]['follow_me_uuid'] = $extension['follow_me_uuid'];
							$array['follow_me'][$x]['follow_me_enabled'] = self::TOGGLE_VALUES[1]; //false
						}
					}

					//increment counter
					$x++;
				}

				//save the changes
				if (is_array($array) && @sizeof($array) != 0) {

					//grant temporary permissions
					$p = new permissions;
					$p->add('extension_edit', 'temp');

					//save the array
					$database->app_name = self::APP_NAME;
					$database->app_uuid = self::APP_UUID;
					$database->save($array);
					unset($array);

					//revoke temporary permissions
					$p->delete('extension_edit', 'temp');

					//send feature event notify to the phone
					if (!empty($_SESSION['device']['feature_sync']['boolean']) && $_SESSION['device']['feature_sync']['boolean'] == "true") {
						foreach ($extensions as $uuid => $extension) {
							$feature_event_notify = new feature_event_notify;
							$feature_event_notify->domain_name = $_SESSION['domain_name'];
							$feature_event_notify->extension = $extension['extension'];
							$feature_event_notify->do_not_disturb = $extension['do_not_disturb'];
							$feature_event_notify->ring_count = ceil($extension['call_timeout'] / 6);
							$feature_event_notify->forward_all_enabled = $extension['forward_all_enabled'];
							$feature_event_notify->forward_busy_enabled = $extension['forward_busy_enabled'];
							$feature_event_notify->forward_no_answer_enabled = $extension['forward_no_answer_enabled'];
							//workarounds: send 0 as freeswitch doesn't send NOTIFY when destination values are nil
							$feature_event_notify->forward_all_destination = $extension['forward_all_destination'] != '' ? $extension['forward_all_destination'] : '0';
							$feature_event_notify->forward_busy_destination = $extension['forward_busy_destination'] != '' ? $extension['forward_busy_destination'] : '0';
							$feature_event_notify->forward_no_answer_destination = $extension['forward_no_answer_destination'] != '' ? $extension['forward_no_answer_destination'] : '0';
							$feature_event_notify->send_notify();
							unset($feature_event_notify);
						}
					}

					//synchronize configuration
					if (!empty($_SESSION['switch']['extensions']['dir']) && is_readable($_SESSION['switch']['extensions']['dir'])) {
						require_once "app/extensions/resources/classes/extension.php";
						$ext = new extension;
						$ext->xml();
						unset($ext);
					}

					//clear the cache
					$cache = new cache;
					foreach ($extensions as $uuid => $extension) {
						$cache->delete("directory:" . $extension['extension'] . "@" . $_SESSION['domain_name']);
						if ($extension['number_alias'] != '') {
							$cache->delete("directory:" . $extension['number_alias'] . "@" . $_SESSION['domain_name']);
						}
					}

					//set message
					message::add($text['message-toggle']);
				}
				unset($records, $extensions, $extension);
			}
		}

//function
	}

	// class
?>
